src/style-spec/feature_filter/index.js

```diff
@@ -82,10 +84,22 @@ function createFilter(filter: any): FeatureFilter {
     if (compiled.result === 'error') {
         throw new Error(compiled.value.map(err => `${err.key}: ${err.message}`).join(', '));
     } else {
-        return (globalProperties: GlobalProperties, feature: VectorTileFeature) => compiled.value.evaluate(globalProperties, feature);
+        const needGeometry = filterNeedGeometry(filter);
+        return {filter: (globalProperties: GlobalProperties, feature: Feature, canonical?: CanonicalTileID) => compiled.value.evaluate(globalProperties, feature, {}, canonical),
+            needGeometry};
     }
 }
 
+function filterNeedGeometry(filter) {
+    if (filter === true || filter === false) {
+        return false;
+    }
+    if (!Array.isArray(filter) || filter.length === 0) {
+        return false;
+    }
+    return filter[0] === 'within';
```

Comment:
Nit: this could be simplified to `return filter && filter[0] === 'within'` — no need for extensive checks.

Reply:
I remove this checking function. But since the return type from `convertFilter(filter)` is mixed, I found it requires type checking before accessing the array member.